zipline/utils/events.py

```diff
@@ -434,23 +413,28 @@ def date_func(self, dt, env):
         raise NotImplementedError
 
     def calculate_start_and_end(self, dt, env):
-        next_trading_day = _coerce_datetime(
-            env.add_trading_days(
-                self.td_delta,
-                self.date_func(dt, env),
-            )
-        )
+        while True:
+            new_date = self.date_func(dt, env)
+            if not new_date:
+                return
 
-        # If after applying the offset to the start/end day of the week, we get
-        # day in a different week, skip this week and go on to the next
-        while next_trading_day.isocalendar()[1] != dt.isocalendar()[1]:
-            dt += datetime.timedelta(days=7)
-            next_trading_day = _coerce_datetime(
-                env.add_trading_days(
+            try:
+                next_trading_day = env.add_trading_days(
                     self.td_delta,
-                    self.date_func(dt, env),
+                    new_date,
                 )
-            )
+            except NoFurtherDataError:
+                return
+
+            if not next_trading_day:
+                return
```

Comment:
Yeah, you're right. Will have `should_trigger` return False if `next_date_start` is still `None` after it tries to calculate